pkg/cache/cache_test.go

```diff
@@ -69,185 +125,180 @@ var _ = Describe("Informer Cache", func() {
+		})
+
+		It("should be able to list objects by namespace", func() {
+			By("listing pods in test-namespace-1")
+			listObj := &kcorev1.PodList{}
+			Expect(informerCache.List(context.Background(),
+				client.InNamespace(testNamespaceOne),
+				listObj)).NotTo(HaveOccurred())
+
+			By("verifying that the returned pods are in test-namespace-1")
+			Expect(listObj.Items).NotTo(BeEmpty())
+			Expect(listObj.Items).Should(HaveLen(1))
+			actual := listObj.Items[0]
+			Expect(actual.Namespace).To(Equal(testNamespaceOne))
+		})
+
+		It("should deep copy the object unless told otherwise", func() {
+			By("retrieving a specific pod from the cache")
+			out := &kcorev1.Pod{}
+			podKey := client.ObjectKey{Name: "test-pod-2", Namespace: testNamespaceTwo}
+			Expect(informerCache.Get(context.Background(), podKey, out)).NotTo(HaveOccurred())
+
+			By("verifying the retrieved pod is equal to a known pod")
+			Expect(out).To(Equal(knownPod2))
+
+			By("altering a field in the retrieved pod")
+			*out.Spec.ActiveDeadlineSeconds = 4
+
+			By("verifying the pods are no longer equal")
+			Expect(out).NotTo(Equal(knownPod2))
+		})
+
+		It("should error out for missing objects", func() {
+			By("getting a service that does not exists")
+			svc := &kcorev1.Service{}
+			svcKey := client.ObjectKey{Namespace: "unknown", Name: "unknown"}
+
+			By("verifying that an error is returned")
+			err := informerCache.Get(context.Background(), svcKey, svc)
+			Expect(err).To(HaveOccurred())
+			Expect(errors.IsNotFound(err)).To(BeTrue())
+		})
+	})
+
+	Describe("as an Informer", func() {
+		It("should be able to get informer for the object", func(done Done) {
+			By("getting a shared index informer for a pod")
+			pod := &kcorev1.Pod{
+				ObjectMeta: kmetav1.ObjectMeta{
+					Name:      "informer-obj",
+					Namespace: "default",
+				},
+				Spec: kcorev1.PodSpec{
+					Containers: []kcorev1.Container{
+						{
+							Name:  "nginx",
+							Image: "nginx",
+						},
+					},
+				},
+			}
+			sii, err := informerCache.GetInformer(pod)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(sii).NotTo(BeNil())
+			Expect(sii.HasSynced()).To(BeTrue())
+
+			By("adding an event handler listening for object creation which sends the object to a channel")
+			out := make(chan interface{})
+			addFunc := func(obj interface{}) {
+				out <- obj
+			}
+			sii.AddEventHandler(kcache.ResourceEventHandlerFuncs{AddFunc: addFunc})
+
+			By("adding an object")
+			cl, err := client.New(cfg, client.Options{})
+			Expect(err).NotTo(HaveOccurred())
+			Expect(cl.Create(context.Background(), pod)).To(Succeed())
+
+			By("verifying the object is received on the channel")
+			Eventually(out).Should(Receive(Equal(pod)))
+			close(done)
+		})
+
+		It("should be able to get an informer by group/version/kind", func(done Done) {
+			By("getting an shared index informer for gvk = core/v1/pod")
+			gvk := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
```

Comment:
Add a test for when the GVK isn't in the Scheme.  Does the code support getting an informer for an unstructured object?  This should definitely be possible, just not sure if we support it right now.

Reply:
I put a TODO for this test.